server/sonar-server/src/test/java/org/sonar/server/es/queue/RecoveryIndexerTest.java

```diff
@@ -0,0 +1,397 @@
+  }
+
+  @Test
+  public void start_triggers_recovery_run_at_fixed_rate() throws Exception {
+    Settings settings = new MapSettings()
+      .setProperty("sonar.search.recovery.initialDelayInMs", "0")
+      .setProperty("sonar.search.recovery.delayInMs", "1");
+    underTest = spy(new RecoveryIndexer(system2, settings, db.getDbClient(), mock(UserIndexer.class)));
+    AtomicInteger calls = new AtomicInteger(0);
+    doAnswer(invocation -> {
+      calls.incrementAndGet();
+      return null;
+    }).when(underTest).recover();
+
+    underTest.start();
+
+    // wait for 2 runs
+    while (calls.get() < 2) {
+      Thread.sleep(1L);
+    }
+  }
+
+  @Test
+  public void successfully_index_old_records() {
+    EsQueueDto item1 = createUnindexedUser();
+    EsQueueDto item2 = createUnindexedUser();
+
+    ProxyUserIndexer userIndexer = new ProxyUserIndexer();
+    advanceInTime();
+    underTest = newRecoveryIndexer(userIndexer);
+    underTest.recover();
+
+    assertThatQueueHasSize(0);
+    assertThat(userIndexer.called)
+      .extracting(EsQueueDto::getUuid)
+      .containsExactlyInAnyOrder(item1.getUuid(), item2.getUuid());
+
+    assertThatLogsContain(TRACE, "Elasticsearch recovery - processing 2 USER");
+    assertThatLogsContain(INFO, "Elasticsearch recovery - 2 documents processed [0 failures]");
+  }
+
+  @Test
+  public void recent_records_are_not_recovered() {
+    createUnindexedUser();
+    createUnindexedUser();
+
+    ProxyUserIndexer userIndexer = new ProxyUserIndexer();
+    // do not advance in time
+    underTest = newRecoveryIndexer(userIndexer);
+    underTest.recover();
+
+    assertThatQueueHasSize(2);
+    assertThat(userIndexer.called).isEmpty();
+
+    assertThatLogsDoNotContain(TRACE, "Elasticsearch recovery - processing 2 USER");
+    assertThatLogsDoNotContain(INFO, "documents processed");
+  }
+
+  @Test
+  public void do_nothing_if_queue_is_empty() {
+    underTest = newRecoveryIndexer();
+
+    underTest.recover();
+
+    assertThatNoLogsFromRecovery(INFO);
+    assertThatNoLogsFromRecovery(ERROR);
+    assertThatQueueHasSize(0);
+  }
+
+  @Test
+  public void log_exception_on_recovery_failure() {
+    createUnindexedUser();
+    FailingOnceUserIndexer failingOnceUserIndexer = new FailingOnceUserIndexer();
+    advanceInTime();
+
+    underTest = newRecoveryIndexer(failingOnceUserIndexer);
+    underTest.recover();
+
+    // No rows treated
+    assertThatQueueHasSize(1);
+    assertThatLogsContain(ERROR, "Elasticsearch recovery - fail to recover documents");
+  }
+
+  @Test
+  public void scheduler_is_not_stopped_on_failures() throws Exception {
+    createUnindexedUser();
+    advanceInTime();
+    FailingUserIndexer userIndexer = new FailingUserIndexer();
+
+    underTest = newRecoveryIndexer(userIndexer);
+    underTest.start();
+
+    // all runs fail, but they are still scheduled
+    // -> waiting for 2 runs
+    while (userIndexer.called.size() < 2) {
+      Thread.sleep(1L);
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Remove this use of "Thread.sleep()". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2925)